From 5649f57cde82b971461856bd70b6d30d4f1629b8 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Tue, 4 Sep 2018 17:17:53 +0900
Subject: [PATCH 364/909] rcar-vin: rcar-csi2: Fix field detection

commit 4c3fad83e30daa43ace650535f3ab34034029e70 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

The PAL and NTSC signals are bottom first, and the other interlaced
signals are top first in some cases. Therefore, according to the
signal called from the source, this patch modifies the first field
number setting in FLD register.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/rcar-vin/rcar-csi2.c | 20 ++++++++++++++++----
 1 file changed, 16 insertions(+), 4 deletions(-)

diff --git a/drivers/media/platform/rcar-vin/rcar-csi2.c b/drivers/media/platform/rcar-vin/rcar-csi2.c
index dc5ae8025832..f5ba9bf0d99f 100644
--- a/drivers/media/platform/rcar-vin/rcar-csi2.c
+++ b/drivers/media/platform/rcar-vin/rcar-csi2.c
@@ -461,12 +461,13 @@ static int rcsi2_calc_mbps(struct rcar_csi2 *priv, unsigned int bpp)
 	return mbps;
 }
 
-static int rcsi2_start(struct rcar_csi2 *priv)
+static int rcsi2_start(struct rcar_csi2 *priv, struct v4l2_subdev *nextsd)
 {
 	const struct rcar_csi2_format *format;
-	u32 phycnt, vcdt = 0, vcdt2 = 0;
+	u32 phycnt, vcdt = 0, vcdt2 = 0, fld = 0;
 	unsigned int i;
 	int mbps, ret;
+	v4l2_std_id std = 0;
 
 	dev_dbg(priv->dev, "Input size (%ux%u%c)\n",
 		priv->mf.width, priv->mf.height,
@@ -475,6 +476,17 @@ static int rcsi2_start(struct rcar_csi2 *priv)
 	/* Code is validated in set_fmt. */
 	format = rcsi2_code_to_fmt(priv->mf.code);
 
+	ret = v4l2_subdev_call(nextsd, video, g_std, &std);
+	if (ret < 0 && ret != -ENOIOCTLCMD && ret != -ENODEV)
+		return ret;
+
+	if (priv->mf.field != V4L2_FIELD_NONE) {
+		if (std & V4L2_STD_525_60)
+			fld = FLD_FLD_NUM(2);
+		else
+			fld = FLD_FLD_NUM(1);
+	}
+
 	/*
 	 * Enable all Virtual Channels.
 	 *
@@ -508,7 +520,7 @@ static int rcsi2_start(struct rcar_csi2 *priv)
 	rcsi2_write(priv, PHTC_REG, 0);
 
 	/* Configure */
-	rcsi2_write(priv, FLD_REG, FLD_FLD_NUM(2) | FLD_FLD_EN4 |
+	rcsi2_write(priv, FLD_REG, fld | FLD_FLD_EN4 |
 		    FLD_FLD_EN3 | FLD_FLD_EN2 | FLD_FLD_EN);
 	rcsi2_write(priv, VCDT_REG, vcdt);
 	rcsi2_write(priv, VCDT2_REG, vcdt2);
@@ -588,7 +600,7 @@ static int rcsi2_s_stream(struct v4l2_subdev *sd, int enable)
 	if (enable && priv->stream_count == 0) {
 		pm_runtime_get_sync(priv->dev);
 
-		ret = rcsi2_start(priv);
+		ret = rcsi2_start(priv, nextsd);
 		if (ret) {
 			pm_runtime_put(priv->dev);
 			goto out;
-- 
2.17.1

